File: trunk/tools/ControlRecord/ControlRecord_GNUradio/tools/FrameBuffer.py
import numpy as np


################################################################################
class FrameBuffer:

################################################################################
    def __init__(self):
        self.frame_starts = []
        self.frame_cnt = np.array([], dtype=np.uint32)

        # "allocate memory" for buffer
        self.byte_buff = np.zeros(100, dtype=np.uint8) 

        # indicates the number of valid bytes in buffer
        self.byte_buff_len = 0


################################################################################
    def IncreaseBufferSize(self, buff_len = 100) :
        self.byte_buff = np.append( self.byte_buff, np.zeros(buff_len, dtype=np.uint8) )

###########
# alt 1.
#        self.byte_buff = np.resize(self.byte_buff, ( 1, self.byte_buff.size + buff_len ) )[0]


################################################################################
    def ClearFromBeginning(self, buff_len) :

        buff_len = min( buff_len, self.byte_buff_len )

        while len(self.frame_starts) > 0 and self.frame_starts[0] < buff_len :
            self.frame_starts.pop(0)
            self.frame_cnt = np.delete( self.frame_cnt, [0] )
#LINUX compatibility
#            self.frame_starts = self.frame_starts[1:]
#            self.frame_cnt = self.frame_cnt[1:]

        for ii in xrange(len(self.frame_starts)) :
            self.frame_starts[ii] -= buff_len

        self.byte_buff = np.delete( self.byte_buff, xrange(buff_len) )
#LINUX compatibility
#        self.byte_buff[0:self.byte_buff_len - buff_len] = self.byte_buff[buff_len:self.byte_buff_len]

###########
# alt 1.
#        self.byte_buff[:self.byte_buff_len-buff_len] = self.byte_buff[buff_len:self.byte_buff_len]

        self.byte_buff_len -= buff_len
        

################################################################################
    def AddToEnd(self, in_buff) :
        self.byte_buff[self.byte_buff_len : self.byte_buff_len + in_buff.size] = in_buff
        self.byte_buff_len += in_buff.size  

File: trunk/tools/ControlRecord/ControlRecord_GNUradio/GNUradio_blocks/tools/FrameBuffer.cpp
#include "FrameBuffer.h"

#include <stdlib.h>
#include <string.h>


FrameBuffer::FrameBuffer ()
	: byte_buff(NULL),
	byte_buff_len(0),
	byte_buff_total_len(0)
{
}


FrameBuffer::~FrameBuffer () 
{
	if (!byte_buff)
		return;

	free(byte_buff);
	byte_buff = NULL;
}


void FrameBuffer::IncreaseBufferSize( unsigned long inc_buff_len )
{
	byte_buff_total_len += inc_buff_len;
	byte_buff = (unsigned char*) realloc( (void*) byte_buff, byte_buff_total_len );
}


void FrameBuffer::ClearFromBeginning( unsigned long buff_len )
{
	buff_len = buff_len < byte_buff_len ? buff_len : byte_buff_len;

	while ( !frame_starts.empty() && frame_starts.front() < buff_len )
	{
		frame_starts.pop();
		frame_cnt.pop();
	}

	for ( unsigned long i = 0; i < frame_starts.size(); i++ )
	{
		frame_starts.push( frame_starts.front() - buff_len );
		frame_starts.pop();
	}

	memmove( (void*) byte_buff, (void*) (byte_buff + buff_len), byte_buff_len - buff_len );

	byte_buff_len -= buff_len;
}


void FrameBuffer::AddToEnd( unsigned char* in_buff, unsigned long in_buff_len )
{
	unsigned long worst_case_size = byte_buff_len + in_buff_len;
	if ( byte_buff_total_len < worst_case_size )
		IncreaseBufferSize( worst_case_size - byte_buff_total_len );

	memcpy( (void*) (byte_buff + byte_buff_len), (void*) in_buff, in_buff_len );
	byte_buff_len += in_buff_len;
}
